aloha-core/src/main/scala/com/eharmony/aloha/util/rand/Rand.scala

```diff
@@ -0,0 +1,120 @@
+package com.eharmony.aloha.util.rand
+
+/**
+  * Some stateless random sampling utilities.
+  *
+  * @author deaktator
+  * @since 11/3/2017
+  */
+private[aloha] trait Rand {
+
+  type Seed = Long
+  type Index = Int
+
+  /**
+    * Perform the initial scramble.  This should be called '''''once''''' on the initial
+    * seed prior to the first call to `sampleCombination`.
+    * @param seed an initial seed
+    * @return a more scrambled seed.
+    */
+  protected def initSeedScramble(seed: Seed): Seed =
+    (seed ^ 0x5DEECE66DL) & 0xFFFFFFFFFFFFL
+
+
+  /**
+    * Sample a ''k''-combination from a population of ''n''.
+    *
+    * This algorithm uses a linear congruential pseudorandom number generator (see Knuth)
+    * to perform reservoir sampling via "''Algorithm R''".
+    *
+    * It is ~ '''''O'''''('''''n''''').
+    *
+    * If `n` ≤ `k`, then return 0, ..., `n` - 1; otherwise, if `k` < `n`, the
+    * returned array have length `k` with values between 0 and `n - 1` (inclusive)
+    * but it is '''NOT''' guaranteed to be sorted.
+    *
+    * '''NOTE''': This is a pure function.  It produces the same results as if
+    * `java.util.Random` was used to perform reservoir sampling but since it doesn't
+    * carry state, this can be trivially operated in parallel with no locking or CAS
+    * loop overhead.  The consequence is that the `seed` must be provided on every call
+    * and a new seed will be returned as part of the output.
+    *
+    * To get this function to act like `java.util.Random`, the first time it is called, the
+    * seed should be produce by running the desired seed through `initSeedScramble`.  For
+    * instance:
+    *
+    * {{{
+    * val (kComb1, newSeed1) = sampleCombination(4, 2, initSeedScramble(0))
+    * val (kComb2, newSeed2) = sampleCombination(4, 2, newSeed1)
+    * }}}
+    *
+    * For more information, see:
+    *
+    - [[http://luc.devroye.org/chapter_twelve.pdf Non-Uniform Random Variate Generation, Luc Devroye, Ch. 12]]
+    - [[https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_R Reservoir sampling (Wikipedia)]]
+    - [[http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/Random.java grepcode for Random]]
+    - [[https://en.wikipedia.org/wiki/Linear_congruential_generator Linear congruential generator (Wikipedia)]]
+    *
+    * @param n population size
+    * @param k combination size
+    * @param seed the seed to use for random selection
+    * @return a tuple 2 containing the array of 0-based indices representing
+    *         the ''k''-combination and a new random seed.
+    */
+  protected def sampleCombination(n: Int, k: Int, seed: Seed): (Array[Index], Seed) = {
+
+    // NOTE: This isn't idiomatic Scala code but it will operate in hot loops in minimizing
+    //       object creation is important.
+
+    if (n <= k) {
```

Comment:
Honestly, this is very difficult to read, but I'm going to assume that you did your research here judging by the comments.

Reply:
You did say you wanted "*fast*" and that hashing would be too slow.  This is what happens when making deals with the devil.  And the devil is *always* in the details.  *Muah ha ha ha!*